drivers: ethernet: renesas: Support Renesas RA8 Ethernet driver

thaoluonguw opened the pull request:
This PR is to support Ethernet driver for the Renesas RA8 boards
The driver design is based on the using Ethernet and Ethernet PHY FSP driver from hal_renesas to implement.

decsny commented on dts/bindings/ethernet/renesas,ra-ethernet.yaml:
```diff
@@ -0,0 +1,29 @@
+  interrupts:
+    required: true
+
+  phy-type:
```
This is too similar to the generic phy-connection-type. Also I don't think it's a good idea to have an enum of a bunch of different specific phy devices. The phy should be a separate node in DT if it is a separate device in hardware.

duynguyenxa replied:
@decsny , Changed to phy-connection-type

decsny commented on dts/arm/renesas/ra/ra8/ra8x1.dtsi:
```diff
@@ -352,6 +352,16 @@
+			reg = <0x40354100 0xfc>;
+			interrupts = <42 0>;
+			local-mac-address = [00 11 22 33 44 55];
+			phy-type = <RA_PHY_ICS1894>;
```
some info specific external phy doesn't belong in an soc devictree, unless that phy is embedded in the soc

decsny commented on drivers/ethernet/eth_renesas_ra.c:
```diff
@@ -0,0 +1,370 @@
+				}
+			}
+		} else if (res == -EAGAIN) {
+			res = R_ETHER_LinkProcess(&ctx->ctrl);
```
should be using zephyr phy API in the driver, no new vendor hal phy apis please, users hate this

duynguyenxa replied:
@decsny , Thank you for you feedback, we change to implement mdio driver and now is using phy API from Zephyr, could you please help to recheck

duynguyenxa committed:
drivers: mdio: Initial support for renesas RA mdio driver

Add support for mdio driver for Renesas RA MCU series
This support utilize the r_ether_phy driver in hal renesas
to support mdio write and read function

Signed-off-by: Duy Nguyen <duy.nguyen.xa@renesas.com>

duynguyenxa committed:
drivers: eth: phy_mii: Add BMSR second read in update_link state

The ICS1894 phy AN_COMPLETE bit is latched high, this make the
BMSR first read return incorrect status of the AN state, update
one more BMSR read to ensure all latched bit is clear and BMSR
return actual status of the phy chip

Signed-off-by: Duy Nguyen <duy.nguyen.xa@renesas.com>

maass-hamburg commented on drivers/ethernet/Kconfig.renesas_ra:
```diff
@@ -0,0 +1,34 @@
+config ETH_RENESAS_RA
+	bool "Renesas RA Ethernet"
+	default y
+	select USE_RA_FSP_ETHER
```
if your eth driver needs the mdio phy consider adding `select MDIO` or `imply MDIO`, so users don't have to activate it manual.

duynguyenxa replied:
@maass-hamburg It's added

duynguyenxa commented on samples/net/cloud/aws_iot_mqtt/boards/ek_ra8m1.overlay:
```diff
@@ -0,0 +1,41 @@
+		status = "okay";
+	};
+};
+
```
@maass-hamburg , the reason I add overlay for this instead of define on board dts is because for using ethernet on EK RA8M1 board, user need to remove Jumper J61 so Ethernet is not a default on this board original state

maass-hamburg replied:
If it is part of the board it should be part of the boards dts, even if it  is not activated. You can mention that and what is needed to enable it in a comment in the dts and in the boards doc. BY the way if the board docs here on zephyr say "Ethernet support", it is expected that ethernet works without having a overlay.

Or is there some interference with other parts?

duynguyenxa replied:
@maass-hamburg , thank you, I got it. I'll move ethernet node in board dts and add a note in document on how to enable the Ethernet on the board

duynguyenxa committed:
drivers: eth: Initial support for Renesas RA Ethernet driver

This commit is to enable Ethernet drivers support on Renesas RA
MCU, first target support is the Renesas RA8 series

Signed-off-by: Duy Nguyen <duy.nguyen.xa@renesas.com>

duynguyenxa committed:
boards: renesas: ra: Add Ethernet support for RA8 boards

Enable Ethernet controller node and mdio node for RA boards.
Add pinctl for mdio and Ethernet usage

Signed-off-by: Duy Nguyen <duy.nguyen.xa@renesas.com>

duynguyenxa commented on samples/net/cloud/aws_iot_mqtt/boards/mck_ra8t1.conf:
```diff
@@ -0,0 +1,13 @@
+CONFIG_NET_DHCPV4_INITIAL_DELAY_MAX=2
+
+CONFIG_NET_L2_ETHERNET=y
+CONFIG_ETH_DRIVER=y
```
I removed the config

duynguyenxa commented on samples/net/dhcpv4_client/boards/mck_ra8t1.conf:
```diff
@@ -0,0 +1,5 @@
+# Copyright (c) 2024 Renesas Electronics Corporation
+# SPDX-License-Identifier: Apache-2.0
+
+CONFIG_NET_L2_ETHERNET=y
```
@maass-hamburg , I created a Kconfig.defconfig files for EK-RA8M1, EK-RA8D1 and MCK-RA8T1

maass-hamburg commented on samples/net/cloud/aws_iot_mqtt/boards/ek_ra8m1.conf:
```diff
@@ -0,0 +1,7 @@
+# Copyright (c) 2024 Renesas Electronics Corporation
+# SPDX-License-Identifier: Apache-2.0
+
+CONFIG_ENTROPY_DEVICE_RANDOM_GENERATOR=y
```
this is on by default, if the device supports it. so no need to enable it explicit.

duynguyenxa replied:
Thank you, I rebase the change and remove the commit for this.

maass-hamburg commented on samples/net/cloud/aws_iot_mqtt/boards/ek_ra8m1.conf:
```diff
@@ -0,0 +1,7 @@
+CONFIG_ENTROPY_DEVICE_RANDOM_GENERATOR=y
+
+CONFIG_NET_DHCPV4=y
+CONFIG_NET_DHCPV4_INITIAL_DELAY_MAX=2
```
is it necessary to set this to 2 instead using the default of 10?

duynguyenxa replied:
Thank you, I rebase the change and remove the commit for this.

duynguyenxa committed:
sample: net: Add initialize for dup_flag when publish

When publish message, the dup_flag is not intialize
causing it to have random value and making the AWS MQTT
broker to reject the message
Add initialize dup_flag as 0

Signed-off-by: Duy Nguyen <duy.nguyen.xa@renesas.com>

duynguyenxa commented on samples/net/cloud/aws_iot_mqtt/boards/ek_ra8d1.conf:
```diff
@@ -0,0 +1,7 @@
+# Copyright (c) 2024 Renesas Electronics Corporation
+# SPDX-License-Identifier: Apache-2.0
+
+CONFIG_ENTROPY_DEVICE_RANDOM_GENERATOR=y
```
@maass-hamburg , Thank you, I rebase the change and remove the commit for this.

duynguyenxa commented:
@decsny , could you help to recheck this PR

jukkar commented on drivers/ethernet/eth_renesas_ra.c:
```diff
@@ -0,0 +1,495 @@
+
+#define ETHER_ISR_EE_RECEIVE_EVENT_MASK (0x01070000)
+
+#if ((ETHER_TOTAL_BUF_NUM) >= 2)
```
BTW, you could avoid these long list of conditional variable declarations by using `LISTIFY()` macro. It would make the code a bit more readable by avoiding this repetitive code.

duynguyenxa replied:
I changed to use LISTIFY() macro for buffer declare, thank you for your suggestion

decsny commented on drivers/ethernet/phy/phy_mii.c:
```diff
@@ -205,6 +205,14 @@ static int update_link_state(const struct device *dev)
+		/* On some PHY chips, the BMSR bits are latched, so the first read may
+		 * show incorrect status. A second read ensures correct values.
+		 */
+		if (phy_mii_reg_read(dev, MII_BMSR, &bmsr_reg) < 0) {
```
generally don't like to see phy specific code in the generic driver but this is small enough I don't think it's worth making a fuss over

jukkar commented on drivers/ethernet/Kconfig.renesas_ra:
```diff
@@ -0,0 +1,34 @@
+
+config ETH_RENESAS_RA
+	bool "Renesas RA Ethernet"
+	default y
```
This change to `Kconfig.renesas_ra` looks like it is in wrong commit, it should be in Ethernet driver addition commit instead.

duynguyenxa replied:
Thank you for reviewing, I moved this to correct commit

jukkar commented on drivers/ethernet/eth_renesas_ra.c:
```diff
@@ -0,0 +1,415 @@
+
+uint8_t *pp_g_ether0_ether_buffers[ETHER_TOTAL_BUF_NUM] = {
+	LISTIFY(CONFIG_ETH_RENESAS_RX_BUF_NUM, DECLARE_ETHER_RX_BUFFER_PTR, (,)),
+		LISTIFY(CONFIG_ETH_RENESAS_TX_BUF_NUM, DECLARE_ETHER_TX_BUFFER_PTR, (,)) };
```
Indentation off

duynguyenxa replied:
Sorry this was the format when I run clang-format

jukkar commented on drivers/ethernet/eth_renesas_ra.c:
```diff
@@ -0,0 +1,415 @@
+
+	ethernet_init(iface);
+
+	const char *phy_connection_type = DT_INST_PROP_OR(0, phy_connection_type, "rmii");
```
Please introduce variables at the start of the func

duynguyenxa replied:
I moved this to device init function

jukkar commented on drivers/ethernet/eth_renesas_ra.c:
```diff
@@ -0,0 +1,415 @@
+	/* Handle for phy-connection-type is different from default (rmmi) */
+	if (strcmp(phy_connection_type, "rmii") != 0) {
+		if (strcmp(phy_connection_type, "mii") == 0) {
+			/* Configure pins for MII or RMII. Set PHYMODE0 if MII is selected. */
```
The comment looks contradicting here, as if the connection type is "rmii", this block is not entered.
If you want to catch both, then perhaps comparing `if (strstr(phy_connection_type, "mii") != NULL) {` instead?

duynguyenxa replied:
Sorry the implement is a bit confusing, I changed it and also move this to device init function

jukkar commented on drivers/ethernet/eth_renesas_ra.c:
```diff
@@ -0,0 +1,415 @@
+	struct renesas_ra_eth_context *ctx;
+	struct net_pkt *pkt = NULL;
+	uint32_t len = 0;
+	static uint8_t rx_buf[NET_ETH_MAX_FRAME_SIZE] __aligned(4);
```
Is the aligned really needed here, I mean the variables should be aligned properly by the compiler anyway. Of course this works also, but looks unnecessary.

duynguyenxa replied:
It's removed, thank you

jukkar commented on drivers/ethernet/eth_renesas_ra.c:
```diff
@@ -0,0 +1,415 @@
+	}
+
+out:
+	if (!pkt) {
```
We are trying to get rid of these (because of some Misra rule), so it would be better to write this as
`if (pkt == NULL) {` instead.

duynguyenxa replied:
I changed to use if `if (pkt == NULL) {`

maass-hamburg commented on drivers/mdio/mdio_renesas_ra.c:
```diff
@@ -0,0 +1,123 @@
+	return 0;
+}
+
+static const struct mdio_driver_api renesas_ra_mdio_api = {
```
```suggestion
static DEVICE_API(mdio, renesas_ra_mdio_api) = {
```
like here: https://github.com/zephyrproject-rtos/zephyr/pull/82267

duynguyenxa replied:
Thank you, I rebase and change to use the `DEVICE_API`

maass-hamburg commented on west.yml:
```diff
@@ -215,7 +215,7 @@ manifest:
     - name: hal_renesas
       path: modules/hal/renesas
-      revision: 10326518701e25bf336a2eaeb8b5820110e4e6a3
+      revision: pull/34/head
```
this should probably point to the merged commit and not on the PR.

duynguyenxa replied:
I changed it to the commit ID

duynguyenxa committed:
manifest: Update hal_renesas commit ID to support Ethernet

Update hal_renesas new commit ID, in this version there are
support for Ethernet and Ethernet phy in hal layer

Signed-off-by: Duy Nguyen <duy.nguyen.xa@renesas.com>

KhiemNguyenT commented:
@soburi It's in good shape. Please take a look.
